src/main/java/net/snowflake/ingest/streaming/internal/ParquetRowBuffer.java

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright (c) 2022 Snowflake Computing Inc. All rights reserved.
+ */
+
+package net.snowflake.ingest.streaming.internal;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import net.snowflake.client.jdbc.internal.google.common.collect.Sets;
+import net.snowflake.ingest.streaming.OpenChannelRequest;
+import net.snowflake.ingest.utils.Constants;
+import net.snowflake.ingest.utils.Logging;
+import net.snowflake.ingest.utils.Pair;
+import org.apache.parquet.column.ColumnDescriptor;
+import org.apache.parquet.schema.MessageType;
+import org.apache.parquet.schema.PrimitiveType;
+import org.apache.parquet.schema.Type;
+
+/**
+ * The buffer in the Streaming Ingest channel that holds the un-flushed rows, these rows will be
+ * converted to Parquet format for faster processing
+ */
+public class ParquetRowBuffer extends AbstractRowBuffer<ParquetChunkData> {
+  private static final Logging logger = new Logging(ParquetRowBuffer.class);
+
+  private static final String PARQUET_MESSAGE_TYPE_NAME = "bdec";
+
+  private final Map<String, Pair<ColumnMetadata, Integer>> fieldIndex;
+  private final Map<String, String> metadata;
+  private final List<List<Object>> data;
+  private final List<List<Object>> tempData;
+
+  private MessageType schema;
+
+  /**
+   * Construct a ParquetRowBuffer object
+   *
+   * @param channel client channel
+   */
+  ParquetRowBuffer(SnowflakeStreamingIngestChannelInternal<ParquetChunkData> channel) {
+    super(channel);
+    fieldIndex = new HashMap<>();
+    metadata = new HashMap<>();
+    data = new ArrayList<>();
+    tempData = new ArrayList<>();
+  }
+
+  @Override
+  public void setupSchema(List<ColumnMetadata> columns) {
+    fieldIndex.clear();
+    metadata.clear();
+    List<Type> parquetTypes = new ArrayList<>();
+    // Snowflake column id that corresponds to the order in 'columns' received from server
+    // id is required to pack column metadata for the server scanner, e.g. decimal scale and
+    // precision
+    int id = 1;
+    for (ColumnMetadata column : columns) {
```

Comment:
you could do something like `for (int id = 1; id < xx; id++) `instead of declaring id outside

Reply:
true, but I find the other `for(column: columns)` a bit more readable.

Reply:
Yeah, this is probably personal preference, but I feel like having a separate index outside of the for loop seems redundant, and I believe most people will just declare it inline :)